internal/engine/integration/procedure_test.go

```diff
@@ -206,56 +163,198 @@ func Test_Procedures(t *testing.T) {
+			res, err := global.Procedure(ctx, tx, &common.ExecutionData{
+				TransactionData: common.TransactionData{
+					Signer: []byte("test_signer"),
+					Caller: "test_caller",
+					TxID:   "test",
+				},
+				Dataset:   dbid,
+				Procedure: procedureName,
+				Args:      test.inputs,
 			})
+			if test.err != nil {
+				require.Error(t, err)
+				require.ErrorIs(t, err, test.err)
+				return
+			}
 			require.NoError(t, err)
 
-			// get dbid
-			dbs, err := global.ListDatasets([]byte("deployer"))
-			require.NoError(t, err)
-			require.Len(t, dbs, 1)
-			dbid := dbs[0].DBID
-
-			// create initial data
-			for username, posts := range initialData {
-				_, err = global.Procedure(ctx, tx, &common.ExecutionData{
-					TransactionData: common.TransactionData{
-						Signer: []byte("username_signer"),
-						Caller: "username_caller",
-						TxID:   "create_user_" + username,
-					},
-					Dataset:   dbid,
-					Procedure: "create_user",
-					Args:      []any{username},
-				})
-				require.NoError(t, err)
-
-				for i, post := range posts {
-					_, err = global.Procedure(ctx, tx, &common.ExecutionData{
-						TransactionData: common.TransactionData{
-							Signer: []byte("username_signer"),
-							Caller: "username_caller",
-							TxID:   "create_post_" + username + "_" + fmt.Sprint(i),
-						},
-						Dataset:   dbid,
-						Procedure: "create_post",
-						Args:      []any{username, post},
-					})
-					require.NoError(t, err)
+			require.Len(t, res.Rows, len(test.outputs))
+
+			for i, output := range test.outputs {
+				require.Len(t, res.Rows[i], len(output))
+				for j, val := range output {
+					require.Equal(t, val, res.Rows[i][j])
 				}
 			}
+		})
+	}
+}
 
-			// parse out procedure name
-			procs := strings.Split(test.procedure, " ")
-			procedureName := strings.Split(procs[1], "(")[0]
-			procedureName = strings.TrimSpace(procedureName)
+func Test_ForeignProcedures(t *testing.T) {
+	type testcase struct {
+		name string
+		// foreign is the foreign procedure definition.
+		// It will be deployed in a separate schema.
+		foreign string
+		// otherProc is the procedure that calls the foreign procedure.
+		// It will be included with the foreign procedure.
+		// It should be formattable to allow the caller to format with
+		// the target dbid, and the target procedure should be hardcoded.
+		otherProc string
+		// inputs are the inputs to the test procedure.
+		inputs []any
+		// outputs are the expected outputs from the test procedure.
+		outputs [][]any
+		// if wantErr is not empty, the test will expect an error containing this string.
+		// We use a string, instead go Go's error type, because we are reading errors raised
+		// from Postgres, which are strings.
+		wantErr string
+	}
+
+	tests := []testcase{
+		{
+			name:    "foreign procedure takes nothing, returns nothing",
+			foreign: `foreign procedure do_something()`,
+			otherProc: `procedure call_foreign() public {
+				do_something['%s', 'delete_users']();
+			}`,
```

Comment:
what's the expected result?  expect no err and no outputs?

Reply:
nothing. No error and no outputs, as the procedure should be successful and it does not return anything.